src/backend/distributed/shared_library_init.c

```diff
@@ -997,6 +1026,73 @@ NormalizeWorkerListPath(void)
 }
 
 
+static bool
+NodeConninfoGucCheckHook(char **newval, void **extra, GucSource source)
+{
+	/* this array _must_ be kept in an order usable by bsearch */
+	const char *whitelist[] = {
+		"application_name",
+		"connect_timeout",
+			#if defined(ENABLE_GSS) && defined(ENABLE_SSPI)
+		"gsslib",
+			#endif
+		"keepalives",
+		"keepalives_count",
+		"keepalives_idle",
+		"keepalives_interval",
+			#if defined(ENABLE_GSS) || defined(ENABLE_SSPI)
+		"krbsrvname",
+			#endif
+		"sslcompression",
+		"sslcrl",
+		"sslmode",
+		"sslrootcert"
+	};
+	char *errmsg = NULL;
+	bool conninfoValid = CheckConninfo(*newval, whitelist, lengthof(whitelist), &errmsg);
+
+	if (!conninfoValid)
+	{
+		GUC_check_errdetail("%s", errmsg);
+	}
+
+	return conninfoValid;
+}
+
+
+static void
+NodeConninfoGucAssignHook(const char *newval, void *extra)
+{
+	PQconninfoOption *optionArray = NULL;
+	PQconninfoOption *option = NULL;
+
+	if (newval == NULL)
+	{
+		newval = "";
+	}
+
+	optionArray = PQconninfoParse(newval, NULL);
+	if (optionArray == NULL)
+	{
+		ereport(ERROR, (errmsg("Unexpected")));
+	}
+
+	ResetConnParams();
```

Comment:
Why isn't the `InitConnParams` call sufficient?

If we're thinking of making this function robust to the value changing if we decide to allow this in the future, shouldn't we also reset the hash?

Reply:
Still thinking about this one.

Reply:
I think being robust to changes is in the realm of e.g. changing `.pgpass` or renaming a role while connections are already cached for the active session. Even if we permit these changes they necessarily are going to only take effect in the next session (so this might be something we'd permit with a SIGHUP or something, but would only take effect with new sessions?)

I've added a comment to say "if we ever allow changing this value at runtime, the conninfo cache should also be cleared".

As for "why separate the logic", I just liked having the initialization of the structure and the initial population of its hardcoded values to be separate. I was sort of modeling this after StringInfo in a way (init/reset/add).

Reply:
For some reason Postgres hash tables don't have a reset function, so it might make more sense to have something like:

```c
FreeConnParams();
InitConnParams();
```

Reply:
I believe they do as of PostgreSQL 10 (or maybe I have 11 checked out; I can't remember, but I looked into it last night)…

That said, the reset function frees all the entries, which is problematic if they contain pointers to allocated memory. At minimum I'd want to make a separate child context of `ConnectionContext` I could reset for this purpose and most of this stuff all seems like premature optimization so long as the GUC requires a restart and must be in the conf file.